### PR TITLE
feat(bindable): support aggregate callback

### LINKED ISSUE #2021: [Feature request] ability to aggregate multiple bindable change callbacks
_Originally posted by @ekzobrain in https://github.com/aurelia/aurelia/issues/1957#issuecomment-2132673808_

---

I know about propertyChanged, but that's not what I meant. I meat exactly propertiesChanged(), which might have the following definition:

```typescript
function propertiesChanged(changes: Record<string, { name: string, newValue: unknown, oldValue: unknown }>): void
```

and may be used like this:
```typescript
propertiesChanged({ prop1, prop2 }) {
  if (prop1 && prop2) {
    // both bindables were changed at the same time (in a batch for example). apply first update strategy
  } else if (prop1) {
    // only prop1 was changed - apply second update strategy
  } else {
    // only prop2 was changed - apply third update strategy
  }
}
```

So it will not fire for each changed property, but only once for all batch changes.

---

Some extra comment from @ekzobrain

> There is no way to set a single change callback for multiple properties. In Angular there's an ngOnChanges() callback (https://angular.io/api/core/OnChanges) which fires once for all changed bindables. I would like to have something similar in Aurelia, for example "propertiesChanged()". There are cases when component update strategy heavily depends on wheather one or several related properties changed at once.

### PR DESCRIPTION
**Pull request by bigopon**
## 📖 Description

Currently, as per noted in #2021 by @ekzobrain:

> There is no way to set a single change callback for multiple properties. In Angular there's an ngOnChanges() callback (https://angular.io/api/core/OnChanges) which fires once for all changed bindables. I would like to have something similar in Aurelia, for example "propertiesChanged()". There are cases when component update strategy heavily depends on wheather one or several related properties changed at once.

This is a missing feature of v1 where it was quite difficult to aggregate the changes during 1 flush cycle of a component. Thanks @ekzobrain for bringing this up.

This PR adds support for the ability to declare a aggregated callback on the component instances (attributes/elements) to receive all the changes at once, with the following signature:

```ts
export interface ICustomElementViewModel {
  ...
  propertiesChanged?(changes: Record<string, { newValue: unknown; oldValue: unknown }>): void;
}
```

- Note that if there' multiple writes to the same property, the changes record given to `propertiesChanged` will only have the latest values, i.e:
    ```ts
    class Component {
      prop = 0
    }
    ...
    component.prop = 1;
    ...
    component.prop = 2;
    ```
    Changes record will be:
    ```ts
    {
      prop: { newValue: 2, oldValue: 1 }
    }
    ```
    not:
    ```ts
    {
      prop: { newValue: 2, oldValue: 0 }
    }
    ```
- The timing of the aggregated changes callback will be the immediate next tick, queued via `Promise.resolve().then(...)` Though this may change if it's deemed unsuitable. We will test for a few versions before RC.

### 🎫 Issues

Close #2021

cc @Sayan751 @fkleuver @ekzobrain 

Note: about the timing, we can try to make it synchronous after a `batch`, but I'm not sure if that's desirable.

cc @3cp with this next tick group update, we kind of have a way to deal with state tearing.

### CODE REVIEW

**Comment by ekzobrain** [comment]
Does current timing conform to synchronous binding system? Or it is considered async as an exception?

File: docs/user-docs/components/bindable-properties.md [hunk 1]
```diff
@@ -99,6 +99,43 @@ export class NameComponent {
...
+  } else if (firstName) {
+    // only firstName was changed - apply second update strategy
+    // ...
+  } else {
+    // only lastName was changed - apply third update strategy
+    // ...
+  }
+}
+```
+
```
Comment (Sayan751) [inline docs/user-docs/components/bindable-properties.md:138]: I think that it needs to be documented, how the callbacks will be invoked when all of them `firstNameChanged`, `propertyChanged`, and `propertiesChanged`. This is already tested, but missing in docs.

**Comment by 3cp** [comment]
From reading I don't think this handles deep change? That was the source of my state-tearing of aurelia dnd service.

```js
class VM {
  someService // Injected service, might be 3rd party
}

class SomeService {
  mutate() {
    this.a = ...;
    this.b = ...;
  }
}
```

**Comment by bigopon** [comment]
> Does current timing conform to synchronous binding system? Or it is considered async as an exception?

It is an exception, everything is synchronous except this aggregated one. Maybe it should be made clearer in the doc.

> Also why not use PropertyKey instead of string in propertiesChanged() signature just as in propertyChanged?

I originally had more than `string`, but then realised we don't support anything other than `string` for `@bindable`, since you wouldn't be able to use it in the template anyway, so changed it to `string`. `propertyChanged` should be updated to reflect this too.

**Comment by bigopon** [comment]
> From reading I don't think this handles deep change? That was the source of my state-tearing of aurelia dnd service.

Indeed, this is only for `@bindable` properties. It won't handle deep mutation. Can you please help note how the observation & mutation are done when state tearing happens, at least for the awareness?

**Comment by 3cp** [comment]
Like this:
```js
class VM
  service: SomeService;
  
  // Au1 will get true all the time
  // Au2 will have a false spike between a and b assignments.
  get computedState() {
    return this.service.a === this.service.b;
  }
}

class SomeService {
  a = "1";
  b = "1";

  someMethod() {
    this.a = "2";
    // Au2 introduced side effects between this two lines.
    // This is way against intuition.
    //
    // Note "SomeService" can be 3rd party lib that
    // User has no control of.
    //
    // I don't think this is a good practice.
    this.b = "2";
  }
}
```

**Comment by 3cp** [comment]
As I repeatedly said, au1 has no issue with 3rd party lib.

No matter what you fix in au2 sync mode, it only applies to the code that user can modify, not 3rd party lib.

It's not humble to introduce side effects to 3rd party lib.

File: docs/user-docs/components/bindable-properties.md (hunk 1 above)
Comment (bigopon) [reply docs/user-docs/components/bindable-properties.md:138]: thanks, added the doc for this 👍

**Comment by bigopon** [comment]
Thanks @3cp , I've created a separate RFC at #2025 to deal with the timing, this PR is good to go without having to wait for the finalisation of that RFC.

**Pull request merged**